src/mca/gds/base/gds_base_fns.c

```diff
@@ -127,23 +129,70 @@ pmix_status_t pmix_gds_base_store_modex(struct pmix_namespace_t *nspace,
         }
         if (PMIX_SUCCESS != rc) {
             /* we have an error */
+            PMIX_ERROR_LOG(rc);
             PMIX_DESTRUCT(&bkt);
-            goto error;
+            goto exit;
         }
 
         // Check that this blob was accumulated with the same data collection setting
         if (have_ctype) {
             if (ctype != (pmix_collect_t)byte) {
                 rc = PMIX_ERR_INVALID_ARG;
                 pbkt.base_ptr = NULL;
-                goto error;
+                PMIX_ERROR_LOG(rc);
+                PMIX_DESTRUCT(&bkt);
+                goto exit;
             }
         }
         else {
             ctype = (pmix_collect_t)byte;
             have_ctype = true;
         }
 
+        /* unpack the size of uniq keys names in the map */
+        cnt = 1;
+        PMIX_BFROPS_UNPACK(rc, pmix_globals.mypeer,
+                           &bkt, &kmap_size, &cnt, PMIX_UINT32);
+        if (PMIX_ERR_UNPACK_READ_PAST_END_OF_BUFFER == rc) {
+            rc = PMIX_SUCCESS;
+            PMIX_DESTRUCT(&bkt);
+            break;
+        } else if (PMIX_SUCCESS != rc) {
+            PMIX_ERROR_LOG(rc);
+            PMIX_DESTRUCT(&bkt);
+            break;
+        }
+
+        /* init and unpack key names map, the position of the key name
+         * in the array determines the unique key index */
+        kmap = (char**)(calloc(kmap_size + 1, sizeof(char*)));
+        if (NULL == kmap) {
+            rc = PMIX_ERR_OUT_OF_RESOURCE;
+            PMIX_ERROR_LOG(rc);
+            goto exit;
+        }
+        cnt = kmap_size;
+        PMIX_BFROPS_UNPACK(rc, pmix_globals.mypeer, &bkt,
+                           kmap, &cnt, PMIX_STRING);
+        if (PMIX_SUCCESS != rc) {
+            PMIX_ERROR_LOG(rc);
+            PMIX_DESTRUCT(&bkt);
+            /* do not use `pmix_argv_free` here since `kmap`
+             * allocated preventively with the supposed size
+             * so pmix_argv doesn't know the size of the array */
+            free(kmap);
```

Comment:
Actually, since you allocated kmap with "calloc", you can and should use pmix_argv_free here. It will simply free all strings in the array until it hits NULL, which calloc guarantees by having initialized all kmap entries to zero.

Reply:
Fixed